src/ui/carousel/carousel.ts

```diff
@@ -0,0 +1,506 @@
+  private readonly activeCssClassSet: Set<string>;
+  private readonly beforeCssClass: string;
+  private readonly beforeCssClassMap: DefaultMap<number, Set<string>>;
+  private readonly afterCssClass: string;
+  private readonly afterCssClassMap: DefaultMap<number, Set<string>>;
+  private readonly condition: () => boolean;
+  private readonly container: HTMLElement;
+  private readonly slides: HTMLElement[];
+  private readonly transition: Transition;
+  private readonly allowLooping: boolean;
+  private readonly onTransitionCallbacks: Array<(carousel: Carousel) => void>;
+  private readonly onDisposeCallbacks: Array<(carousel: Carousel) => void>;
+  private readonly slideCssClasses: DefaultMap<HTMLElement, Set<string>>;
+  private readonly raf: Raf;
+  private readonly synchronizer: CarouselSynchronizer;
+  private transitionTarget: TransitionTarget;
+  private lastActiveSlide: HTMLElement;
+
+  /**
+   * @param container Parent element of slides.
+   * Element that acts as the carousel, in which all slides are contained.
+   * Does not need to be the direct parent.
+   *
+   * @param slides HTMLElements containing slide content.
+   *
+   * @param condition Under what conditions the carousel should run
+   * @param onTransitionCallbacks Functions run when the active slide changes.
+   * @param onDisposeCallbacks Functions run when the carousel is disposeed.
+   * @param activeCssClass Class to apply to active slide.
+   * @param beforeCssClass Class to apply to slides before active slide.
+   * @param afterCssClass Class to apply to slides after active slide.
+   * @param allowLooping Whether the carousel should be allowed to loop.
+   * @param transition How the Carousel should transition.
+   * Please see the transitions folder inside the carousel folder for options.
+   */
+  constructor(
+      container: HTMLElement,
+      slides: HTMLElement[],
+      {
+        condition = () => true,
+        onTransitionCallbacks = [],
+        onDisposeCallbacks = [],
+        activeCssClass = DefaultCssClass.ACTIVE_SLIDE,
+        beforeCssClass = DefaultCssClass.BEFORE_SLIDE,
+        afterCssClass = DefaultCssClass.AFTER_SLIDE,
+        allowLooping = true,
+        transition = null
+      }: CarouselOptions = {}
+  ) {
+    if (slides.length < 1) {
+      throw new Error('Cannot start carousel without slides');
+    }
+    this.raf = new Raf(() => this.loop());
+    this.activeCssClass = activeCssClass;
+    this.beforeCssClass = beforeCssClass;
+    this.afterCssClass = afterCssClass;
+    this.allowLooping = allowLooping;
+    this.condition = condition;
+    this.container = container;
+    this.lastActiveSlide = null;
+    this.onTransitionCallbacks = onTransitionCallbacks;
+    this.onDisposeCallbacks = onDisposeCallbacks;
+    this.slides = slides;
+    this.transition = transition !== null ? transition : new CssClassesOnly();
+    this.transitionTarget = null;
+    this.synchronizer = CarouselSynchronizer.getSingleton(this);
```

Comment:
I think overall the pattern of components looking for instances of singletons can be confusing to follow and makes architecture a harder to follow.

In general, I find it easier if components are self composed and follow a top down composition.  

Here you have a case of Carousel, doing horizontal lookup of CarouselSynchronizer and then later calling methods of CarouselSynchronizer.  I'm not exactly sure what sync is doing here but it feels a bit coupled.

Can we architect this in a different way?

I think instead this might be easier to follow if the carousel was solely responsible to updating itself.  It would have a render() method.  It should have no knowledge of what a CarouselSynchronize is.  You could have a public method that perhaps disabled the internal Raf like stopLoop().

You then have multiple instances of Carousels.

You instantiate a CarouselSynchronizer([Carousels...]).  CarouselSynchronizer would call stopLoop() on each carousel component and then run it's own raf and run the render loop on each Carousel.

Perhaps this is not exactly what you are doing but the point being, I'm wondering if we can decouple this more.

Reply:
I'll take a look. This has always bothered me as well, one of those things that survived because it was fast to implement and then didn't pose enough problems to warrant priority later.

Reply:
Removed reference to CarouselSynchronizer from within carousel.ts.